kernel/kernel-api/src/main/java/io/delta/kernel/internal/ScanImpl.java

```diff
@@ -309,4 +346,79 @@ private CloseableIterator<FilteredColumnarBatch> applyDataSkipping(
               filteredScanFileBatch.getData(), Optional.of(newSelectionVector));
         });
   }
+
+  /**
+   * Wraps a scan file iterator such that we emit {@link ScanReport} to the engine upon success and
+   * failure. Since most of our scan building code is lazily executed (since it occurs as
+   * maps/filters over an iterator) potential errors don't occur just within `getScanFile`s
+   * execution, but rather may occur as the returned iterator is consumed. Similarly, we cannot
+   * report a successful scan until the iterator has been fully consumed and the log read/filtered
+   * etc. This means we cannot report the successful scan within `getScanFiles` but rather must
+   * report after the iterator has been consumed.
+   *
+   * <p>This method takes a scan file iterator. It wraps the methods `next` and `hasNext` such that
+   * in the case of an exception originating from there, we will first report a failed ScanReport
+   * before propagating the exception. It also reports a success ScanReport when the iterator is
+   * closed, if and only if, all the elements have been consumed.
+   */
+  private CloseableIterator<FilteredColumnarBatch> wrapWithMetricsReporting(
+      CloseableIterator<FilteredColumnarBatch> scanIter, ScanReportReporter reporter) {
+    return new CloseableIterator<FilteredColumnarBatch>() {
+
+      @Override
+      public void close() throws IOException {
+        try {
+          if (!scanIter.hasNext()) {
+            // The entire scan file iterator has been successfully consumed, report successful Scan
+            reporter.reportSuccess();
+          } else {
+            // TODO what should we do if the iterator is not fully consumed?
+            //  - Do nothing (we don't know why the iterator is being closed early or what is going
+            //    on in the connector)
+            //  - Report a failure report with some sort of placeholder exception
+            //  - Something else?
```

Comment:
TODO - decide on this

For example, this is possible if there is an error somewhere within the connector code; they close the iterator early but Kernel will never see an exception. However, it's possible this happens for other reasons, we have no way of knowing why the connector is closing the iterator before consuming all the actions.

Reply:
Additional option would be reporting whether a Scan is "complete" or not

Reply:
but when would we ever be able to invoke
```java
// okay, now I know for sure that the iterator was NOT fully consumed
scanReport = ... something ... maybe that says it wasn't fully consumed ...
engine.getMetricsReporters().forEach(reporter -> reporter.report(scanReport));
```

I don't think we can?

Reply:
When the iterator is closed (inside the close method) but `hasNext = true`. This is for when the iterator is not fully consumed but the connector is correctly closing it.

Nothing we can do if it is not closed.

Reply:
Thanks for explaining! Is it much work to report a success but also indicate that it was not fully consumed? A usecase here could be LIMIT, right? An engine would consume the AddFiles until some LIMIT number of rows, and then close the iterator without fully consuming it. Would be nice to capture + support this.

Reply:
I think that's a valid example of a successful case where this might happen. However that's likely also indistinguishable from the case where an error is thrown within the connector, the scan is interrupted/not completed and the iterator is closed. I'm not sure how we want to address these possibilities.

I think maybe this would be this option I called out
> Additional option would be reporting whether a Scan is "complete" or not

We can add a field such as `boolean isFullyConsumed`; how this is interpreted can be left for later (e.g. wrt to the two cases outlined above).

Reply:
I think using the boolean is good enough! Connectors are responsible to know whether or not an exception was thrown inside their own connector.